woocommerce/woocommerce-ios
# Order details: add-ons from `_pao_ids` metadata - Yosemite & app layer changes

Part of #10419 


## Why


Currently, we're getting the product add-ons of an order line item by manually mapping the metadata entries with the product and global add-ons that need to be synced separately. Starting Product Add-Ons v5.0.2 pe5pgL-3wp-p2#comment-2967, a new metadata field `_pao_ids` became available to get a list of add-ons of an order line item. After the storage and networking layer changes, this PR updated the Yosemite & app layers to persist the add-ons from each order item and then use them in the app.

I decided to keep the backward compatibility since it wasn't too much work. In the add-ons use cases, it returns `OrderItem.addOns` if it is non-empty, then if it's empty it falls back to the original `AddOnCrossreferenceUseCase` to check on the order item attributes.

## How

🗒️ There are a lot of files with code changes because of the `OrderItem` initialization especially in test cases. We can consider replacing them with `Fakes` & `Copiable` so that any new parameters wouldn't result in lots of file changes.
🗒️ After the PR, the add-on values in order details are still from the attributes because there are a bunch of changes to update in `ProductDetailsCellViewModel`. I decided to minimize the diffs in this PR and will make the changes in the next PR to update the design https://github.com/woocommerce/woocommerce-ios/issues/10419

- Added `addOns` to `AggregateOrderItem` from `OrderItem.addOns`, since most of the UI uses `AggregateOrderItem` after combining with refunded order items in `AggregateDataHelper`
- Currently, there are two screens where the add-ons are displayed for an order. The logic has been updated to get add-ons from `OrderItem.addOns` with a fallback to attributes using `AddOnCrossreferenceUseCase`:
  - Order details: changes are in `OrderDetailsDataSource`
  - Order details > mark as complete (review order): changes are in `ReviewOrderViewModel` and `ReviewOrderViewController`
- To differentiate add-ons from other order item attributes like variation attributes, `AddOnCrossreferenceUseCase` now returns an array of add-ons instead of attributes. This would be helpful in the https://github.com/woocommerce/woocommerce-ios/issues/10419 as we only want to change the design for add-ons

## Testing instructions


Prerequisite: the store has Product Add-Ons extension, and at least an order with different types of add-ons

- Log in to the store in the prerequisite if needed
- In Settings > Experimental Features, enable the `view add-ons` feature
- Go to the Orders tab
- Tap on an order with different types of add-ons --> the add-on values should match the values in core
- Tap `View add-ons` --> the add-ons should match the values in core
- Go back to the order details screen
- If `Mark Order Complete` CTA isn't available, mark the order as paid and then update the order status to `Processing`
- Tap on `Mark Order Complete` CTA --> the review order screen should show the add-on values that match core
- Tap `View add-ons` --> the add-ons should match the values in core

---
- [ ] @jaclync tests that disabling `view add-ons` in Settings > Experimental Features hides the CTA in order details and review order screens

## Screenshots


screen | order details | add-ons
-- | -- | --
order details | ![Simulator Screenshot - iPhone 14 Pro - 2023-09-01 at 16 49 16](https://github.com/woocommerce/woocommerce-ios/assets/1945542/a3371b8e-2887-4f87-aaf5-8518bbecb221) | ![Simulator Screenshot - iPhone 14 Pro - 2023-09-01 at 16 49 20](https://github.com/woocommerce/woocommerce-ios/assets/1945542/8494b7da-695f-487b-8391-015b797d1787)
review order | ![Simulator Screenshot - iPhone 14 Pro - 2023-09-01 at 16 49 29](https://github.com/woocommerce/woocommerce-ios/assets/1945542/c5a7ef7a-7eef-4c23-8051-2a22b93de578) | ![Simulator Screenshot - iPhone 14 Pro - 2023-09-01 at 16 49 31](https://github.com/woocommerce/woocommerce-ios/assets/1945542/47d395ec-994a-42dd-a887-55900092ff02)


---
